site/docs/design-docs/012-improved-error-handling.md

```diff
@@ -0,0 +1,113 @@
+---
+title: 012 - Improved Error Tolerance of Streaming Jobs
+description: Make stream jobs capable of surviving errors with fixable causes.
+---
+
+*Target Release*: 4.3
+
+## Goal
+
+Since streaming jobs can't simply be re-run, like batch ones, having
+them fail can result in significant loss. We want them to be as
+resilient as possible and only ever really fail due to the most severe
+of causes.
+
+Right now any kind of failure, like exceptions in user code or thrown
+by sources or sinks will stop the execution of a job. Moreover, not only
+does execution stop, snapshots of the job are also lost, so even if
+the root problem can be fixed, there is no way to recover or resume
+the failed job.
+
+We want to improve on this by only suspending jobs with such failures
+and preserving their snapshots.
+
+## Breaking-change
+
+It might be argued that suspending a job on failure, instead of letting
+it fail completely is a breaking-change, as far as behaviour is
+concerned.
+
+One might also argue that it's broken behaviour which just took a long
+time to fix.
+
+Anyways, to be safe we might preserve the current behaviour as default
+and make the suggested changes optional. Maybe as a new element in
+`JobConfig`, called `suspend_on_failure`, disabled unless otherwise
+specified.
+
+## Notify client
+
+When we suspend a job due to a failure we need to notify the client that
+submitted it and give enough information to facilitate the repair of the
+underlying root cause.
+
+For this reason we should introduce a new `JobStatus`, called
+`SUSPENDED_DUE_TO_ERROR` and provide details in the `Exception` set in
+the `CompletableFuture` returned by `job.getFuture()` (also used by
+`job.join()`).
```

Comment:
Hmm... ok, but that's a bit shady... Since `join` is pretty much irrelevant for streaming jobs, how about we introduce another similar operation, which would be more useful? Something like `Job.track()` or `Job.monitor()`, which would return when the streaming job is suspended? (Probably the choice of names is terrible, but you get the idea.) Seems cleaner to me.